druid/src/widget/container.rs

```diff
@@ -114,25 +73,46 @@ impl<T: Data> Widget<T> for ContainerRaw<T> {
 
         // Paint border
         if let Some(ref border) = self.style.border {
-            let rect = Rect::from_origin_size((0.0, 0.0), base_state.size());
+            let offset = border.width / 2.0;
+            let size = Size::new(
+                base_state.size().width - border.width,
+                base_state.size().height - border.width,
+            );
+            let rect = Rect::from_origin_size((offset, offset), size);
             paint_ctx
                 .render_ctx
                 .stroke(rect, &border.brush, border.width);
         }
 
         // Paint child
-        self.inner.paint(paint_ctx, base_state, data, env);
+        self.inner.paint_with_offset(paint_ctx, data, env);
     }
 
     fn layout(&mut self, ctx: &mut LayoutCtx, bc: &BoxConstraints, data: &T, env: &Env) -> Size {
-        self.inner.layout(ctx, bc, data, env)
+        bc.debug_check("Container");
+
+        // Shrink constraints by border offset
+        let border_width = match self.style.border {
```

Comment:
alternatively,

```rust
self.style.border.as_ref().map(|b| b.width).unwrap_or_default();
```

(not necessary to change, just wanted to point out the one-liner)